plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractSemanticRegionsFinder.java

```diff
@@ -0,0 +1,414 @@
+		return findFirst(createPredicate(assignment));
+	}
+
+	@Override
+	public List<ISemanticRegion> assignments(Assignment... assignments) {
+		return findAll(createPredicate(assignments));
+	}
+
+	protected void collectMatchableElements(AbstractElement ele, Collection<AbstractElement> result) {
+		switch (ele.eClass().getClassifierID()) {
+			case XtextPackage.RULE_CALL:
+				assertNoEObjectRuleCall((RuleCall) ele);
+			case XtextPackage.KEYWORD:
+				result.add(ele);
+				break;
+			case XtextPackage.CROSS_REFERENCE:
+				collectMatchableElements(((CrossReference) ele).getTerminal(), result);
+				break;
+			case XtextPackage.ASSIGNMENT:
+				collectMatchableElements(((Assignment) ele).getTerminal(), result);
+				break;
+			case XtextPackage.ALTERNATIVES:
+			case XtextPackage.GROUP:
+			case XtextPackage.UNORDERED_GROUP:
+				for (AbstractElement child : ((CompoundElement) ele).getElements())
+					collectMatchableElements(child, result);
+				break;
+		}
+	}
+
+	protected Predicate<ISemanticRegion> createPredicate(AbstractElement ele) {
+		switch (ele.eClass().getClassifierID()) {
+			case XtextPackage.RULE_CALL:
+				assertNoEObjectRuleCall((RuleCall) ele);
+			case XtextPackage.KEYWORD:
+				return new GrammarElementPredicate(ele);
+		}
+		return createPredicate(new AbstractElement[] { ele });
+	}
+
+	protected Predicate<ISemanticRegion> createPredicate(AbstractElement... ele) {
+		Set<AbstractElement> result = Sets.newHashSet();
+		for (AbstractElement e : ele)
+			collectMatchableElements(e, result);
+		switch (result.size()) {
+			case 0:
+				return Predicates.alwaysFalse();
+			case 1:
+				return new GrammarElementPredicate(result.iterator().next());
+			default:
+				return new GrammarElementsPredicate(result);
+		}
+	}
+
+	@Override
+	public ISemanticRegion crossRef(CrossReference crossReference) {
+		return findFirst(createPredicate(crossReference));
+	}
+
+	@Override
+	public List<ISemanticRegion> crossRefs(CrossReference... crossReferences) {
+		return findAll(createPredicate(crossReferences));
+	}
+
+	@Override
+	public ISemanticRegion element(AbstractElement element) {
+		return findFirst(createPredicate(element));
+	}
+
+	@Override
+	public List<ISemanticRegion> elements(AbstractElement... elements) {
+		return findAll(createPredicate(elements));
+	}
+
+	@Override
+	public ISemanticRegion feature(EStructuralFeature feature) {
+		assertNoContainment(feature);
+		return findFirst(new FeaturePredicate(feature));
+	}
+
+	@Override
+	public List<ISemanticRegion> features(EStructuralFeature... features) {
+		Set<Predicate<ISemanticRegion>> predicates = Sets.newHashSet();
+		for (int i = 0; i < features.length; i++) {
+			EStructuralFeature feat = features[i];
+			assertNoContainment(feat);
+			predicates.add(new FeaturePredicate(feat));
+		}
+		return findAll(Predicates.or(predicates));
+	}
+
+	protected ImmutableList<ISemanticRegion> findAll(Predicate<ISemanticRegion> predicate) {
```

Comment:
refactored this in 1d942e3 to be an abstract method. Now subclasses get to decide on an implementation explicitly.